staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,178 @@ func TestForgetWatcher(t *testing.T) {
 	assertCacherInternalState(0, 0)
 	require.Equal(t, 2, forgetCounter)
 }
+
+// TestGetWatchCacheResourceVersion test the following cases:
+//
+// +-----------------+---------------------+-----------------------+
+// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+// +=================+=====================+=======================+
+// | Unset           | true/false          | nil/true/false        |
+// | 0               | true/false          | nil/true/false        |
+// | 105             | true/false          | nil/true/false        |
+// +-----------------+---------------------+-----------------------+
+// where:
+// - false indicates the value of the param was set to "false" by a test case
+// - true  indicates the value of the param was set to "true" by a test case
+func TestGetWatchCacheResourceVersion(t *testing.T) {
+	listOptions := func(allowBookmarks bool, sendInitialEvents *bool, rv string) storage.ListOptions {
+		p := storage.Everything
+		p.AllowWatchBookmarks = true
+
+		opts := storage.ListOptions{}
+		opts.Predicate = p
+		opts.SendInitialEvents = sendInitialEvents
+		opts.ResourceVersion = rv
+		return opts
+	}
+
+	scenarios := []struct {
+		name                       string
+		opts                       storage.ListOptions
+		parsedWatchResourceVersion int
+
+		expectedWatchResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name:                         "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedRV=parsedWatchResourceVersion",
+			opts:                         listOptions(true, nil, ""),
+			parsedWatchResourceVersion:   6,
```

Comment:
is this clear ? I think this is the only way to test that the `parsedWatchResourceVersion` is returned.

Reply:
No - I'm less interested in checking that parseResourceVersion was returned - I'm rather interested in the exact value.

So we should pass what will be passed in real code (which in case of RV=unset would be 0, right?

Reply:
correct, i decided to pass a fake `parsedWatchResourceVersion ` to distinguish `RV=0` from `RV=unset` as you noticed that both cases in real code will be `0`.

Reply:
The function itself is distinguishing it based on len(options.ResourceVersion).

Please really use the data that will really be used in production - so 0.

Reply:
You can distinguish the result by having the RV returned by underlying storage to be different the RV passed in - that's what I really want to check.

Reply:
> You can distinguish the result by having the RV returned by underlying storage to be different the RV passed in

We have that for the following cases:
- `RV=unset, allowWatchBookmarks=true, sendInitialEvents=true, expectedRV=100(storage)`
- `RV=unset, allowWatchBookmarks=false, sendInitialEvents=true, expectedRV=100(storage), legacy`

Reply:
I don't follow - you should pass 0 in those cases (as parsedResourceVersion) and we will verify the result.

Reply:
pushed, ptal.